### DISCUSSION TITLE
Referencing Nested Entries in Rich Text field from Contentful

### DISCUSSION THREAD

**Discussion by stevenyang1011** [gatsbyjs/gatsby#31932] [Help] [answered]
Hi. I have a problem with embedding multi levels of entries in the rich text field on Contentful with the gatsby-source-contentful plugin. For example in the Rich text field of a page content model I want to embed a Grid entry:

```
fragment ContentfulGridFragment on ContentfulGrid{
    title
    contentful_id
    columns
    items {
        contentful_id
        content {
            raw
        }
    }
  }
```

And say in the Grid there is another rich text field for embedding another Grid entry. The graphQL for this becomes something like:

```
fragment ContentfulGridFragment on ContentfulGrid{
    title
    contentful_id
    columns
    items {
        contentful_id
        content {
            raw
            references{
                ... on ContentfulGrid {
                   title
                  contentful_id
                  columns
                  items {
                      contentful_id
                      content {
                          raw
                      }
                 }
            }
        }
    }
  }
```
I want to enable more flexibility by not having to write the graphQL specifying a limited number of nested levels so that the code can be reused when a new page with more nested grids is created. 

This is only the simplest example of the issue I am having. The graphQL gets more complicated when there are more content models with rich text fields and when I try to allow them being embedded in other rich text fields.

Is there a solution with the current version of the plugin?

Here is a demo of the code I created [https://github.com/stevenyang1011/gatsby-contentful-circular-reference](url)

cc @axe312ger

**Comment by axe312ger** [comment, marked as answer]
This is not related to Contentful, this is more GraphQL in general. I think you have to specify how deep you will go. Maybe the Core team has an idea.

**Reply by wardpeet** [reply]
@stevenyang1011 it's indeed not possible with GraphQL to do a recursive nesting. it's better to export is as a JSON type so you can fetch the whole tree.

**Reply by axe312ger** [reply]
@stevenyang1011 does this answer your question? :)

**Reply by stevenyang1011** [reply]
Thanks guys. Yep I've decided to use an older version of the gatsby-source-contentful plugin which uses JSON for the rich text data.

**Reply by axe312ger** [reply]
@stevenyang1011 that version has massive performance issues because of that JSON field. I'd recommend not to do that and rather write your queries a few layers deep.

**Reply by stevenyang1011** [reply]
@axe312ger Yea I did notice the difference in performance, but it seems like the only solution though.
There are a few models with rich text fields in the actual project, and the editors on Contentful does not have access to the code. So the code has to work if an editor adds another level of embedded entry.

**Reply by axe312ger** [reply]
I wonder which layout/design can allow unlimited nesting of entries with text fields?

Maybe rich text is the wrong solution for your problem?
